src/Particle/SampleStack.cpp

```diff
@@ -88,30 +38,37 @@ void SampleStack::setMaxSamples(int n)
   }
 }
 
-/** save the [first,last) walkers to SampleStack
- */
-void SampleStack::saveEnsemble(walker_iterator first, walker_iterator last)
+MCSample& SampleStack::getSample(unsigned int i) const { return *sample_vector_[i]; }
+
+void SampleStack::saveEnsemble(std::vector<MCSample>& walker_list)
 {
   //safety check
   if (max_samples_ == 0)
     return;
+  auto first = walker_list.begin();
+  auto last  = walker_list.end();
   while ((first != last) && (current_sample_count_ < max_samples_))
   {
-    sample_vector_[current_sample_count_]->put(**first);
+    *sample_vector_[current_sample_count_] = *first;
     ++first;
     ++current_sample_count_;
   }
 }
 
+void SampleStack::appendSample(MCSample walker)
```

Comment:
Can we avoid making copies of walker? It is copied twice. 1st time when passing by value, second time assign to sample_vector_.
```
appendSample(MCSample&& walker)
*sample_vector_[current_sample_count_] = std::move(walker);
```
Does this look better?